docs/design/features/byreflike-generics.md

```diff
@@ -0,0 +1,83 @@
+# Generics parameters of ByRefLike types
+
+Using ByRefLike types in Generic parameters is possible by building upon support added for `ref` fields. Scenarios that would benefit most from this are those involving `Span<T>`. For example, consider the following examples:
+
+- `Span<TypedReference>` &ndash; Represents the general case where a ByRefLike type is used as a Generic parameter. This specific case would be desirable for a more efficient Reflection API.
+- `Span<Span<char>>` &ndash; Nested `Span<T>` types would be of benefit in the parsing result of strings.
+
+## Runtime impact
+
+Supporting ByRefLike type as Generic parameters will impact the following IL instructions:
+
+- `box` &ndash; Types with ByRefLike parameters used in fields cannot be boxed.
+- `stsfld` / `ldsfld` &ndash; Type fields of a ByRefLike parameter cannot be marked `static`.
+- `newarr` / `stelem` / `ldelem` / `ldelema` &ndash; Arrays are not able to contain ByRefLike types.
+    - `newobj` &ndash; For multi-dimensional array construction.
+- `constrained.callvirt` &ndash; If this IL sequence resolves to a method implemented on `object` or default interface method, an error will occur during the attempt to box the instance.
+
+If any of the above instructions are attempted to be used with a ByRefLike type, the runtime will throw an `InvalidProgramException`.
+
+The following instructions are already set up to support this feature since their behavior will fail as currently defined due to the inability to box a ByRefLike type.
+
+- `throw` &ndash; Requires an object reference to be on stack, which can never be a ByRefLike type.
+- `unbox` / `unbox.any` &ndash; Requires an object reference to be on stack, which can never be a ByRefLike type.
```

Comment:
True they require an object reference but the following sequence is effectively how non-boxing casts work in IL: 

```il
        IL_000d: ldarg.0
        IL_000e: box !!T
        IL_0013: isinst [System.Private.CoreLib]System.Int32
        IL_0018: unbox.any [System.Private.CoreLib]System.Int32
        IL_001d: stloc.0
```

This doesn't actually put the `struct` on the heap hence my instinct is it would be valid for `ref struct`. If it's not valid then we effectively lose the ability to do pattern matching and type checks on `ref struct`.

Reply:
The `box` instruction above doesn't actually box the argument?

Reply:
See https://github.com/dotnet/runtime/pull/67129#discussion_r837976315